### PR TITLE
Document elliptic systems

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Adds a protocol to document elliptic systems.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/Protocols/FirstOrderSystem.hpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+ *
+ * - `auxiliary_fields`: A list of tags representing the auxiliary fields, which
+ *   are typically gradients of the primal fields, e.g. \f$v_i = \partial_i u\f$
+ *   for a Poisson equation. These must follow the order of the `primal_fields`.
+ *   Specifically, each auxiliary field must have one rank higher than its
+ *   corresponding primal field.
+ *
+ * - `primal_fluxes`: A list of tags representing the primal fluxes
+ *   \f$F_{u_A}^i\f$. These are typically some linear combination of the
+ *   auxiliary fields with raised indices, e.g. \f$v^i = g^{ij}v_j\f$ for a
```
Comment (nilsdeppe) [inline src/Elliptic/Protocols/FirstOrderSystem.hpp:105]: could you define `g^{ij}` please? In the evolution code we are trying to use `g^{ij}` as the spatial part of the spacetime metric, not the spatial metric

File: src/Elliptic/Protocols/FirstOrderSystem.hpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+ *   first index because their divergence defines the auxiliary fields.
+ *
+ * - `background_fields`: A list of tags representing the variable-independent
+ *   background fields in the equations. Examples are a background metric,
+ *   associated fixed geometry quantities such as Christoffel symbols or the
+ *   Ricci scalar, or any other fixed field that determines the problem to be
+ *   solved such as matter sources in the Einstein constraint equations.
+ *
+ * - `inv_metric_tag`: The tag that defines the background geometry. This is the
+ *   metric responsible for normalizing one-forms, such as face normals.
```
Comment (nilsdeppe) [inline src/Elliptic/Protocols/FirstOrderSystem.hpp:123]: can you clarify if this should be the spatial metric or the spatial part of the spacetime metric?

File: src/Elliptic/Protocols/FirstOrderSystem.hpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+ *   2. The `argument_tags`
+ *   3. The `auxiliary_fields`
+ *
+ *   The second `apply` function takes these arguments in this order:
+ *
+ *   1. The `auxiliary_fluxes` as not-null pointer
+ *   2. The `argument_tags`
+ *   3. The `primal_fields`
+ *
+ *   The functions can assume the output buffers are already correctly sized.
```
Comment (nilsdeppe) [inline src/Elliptic/Protocols/FirstOrderSystem.hpp]: is there any guarantee on the value at input? E.g. will they be zeroed?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
You can squash immediately

File: src/Elliptic/Protocols/FirstOrderSystem.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:105]: this is just a curved-space Poisson equation so there is no time. `g^{ij}` is just the metric of the Riemannian manifold that the Poisson eqn is formulated on. I added some docs

File: src/Elliptic/Protocols/FirstOrderSystem.hpp (hunk 2 above)
Comment (nilsvu) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:123]: It's up to the system what this metric is :) For a curved-space Poisson eqn it's just the background metric. For the XCTS equations it's the conformal metric. I think I can't really make it more specific at this point than "the geometry that the equations are formulated on".

File: src/Elliptic/Protocols/FirstOrderSystem.hpp (hunk 3 above)
Comment (nilsvu) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp]: there is not, I added a comment 👍

File: src/Elliptic/Protocols/FirstOrderSystem.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:105]: so it's the spatial metric? In that case would it be better to use `gamma^{ij}`?

File: src/Elliptic/Protocols/FirstOrderSystem.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:123]: Is it not the spatial metric of the background geometry? I'm asking whether it is a spatial metric, spatial components of a spacetime metric, or the spacetime metric. At least in the evolution code this decision is made regardless of what the background geometry is (we use the inverse spatial metric)

Comment (nilsvu) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:123]: There is only one metric in the elliptic equations, and that's the one coupling the principal part of the PDEs. It's just "the background metric g_ij" for a curved-space Poisson equation, and it's "the conformal metric \bar{gamma}_ij" for the XCTS equations. For the XCTS equations the conformal metric is used to assemble the spacetime metric, by combining it with the conformal factor, lapse, shift etc.

Comment (nilsdeppe) [reply src/Elliptic/Protocols/FirstOrderSystem.hpp:123]: okay, thanks :)

**Review by nilsdeppe — DISMISSED** [review]
Sorry I didn't notice your updates on this. I'm happy with the updates, but it needs a rebase to deal with a conflict in a CMake file

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer looks like this needs a rebase again... Sorry I didn't merge before the conflict.

**Comment by nilsvu** [comment]
@nilsdeppe this is rebased now.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**